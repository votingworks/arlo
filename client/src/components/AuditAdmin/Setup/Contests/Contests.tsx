/* eslint-disable jsx-a11y/label-has-associated-control */
import React from 'react'
import ContestForm from './ContestForm'
import ContestSelect from './ContestSelect'
import { IAuditSettings } from '../../../useAuditSettings'

export interface IContestsProps {
  electionId: string
  auditType: IAuditSettings['auditType']
  isTargeted: boolean
  goToNextStage: () => void
  goToPrevStage: () => void
}

const Contests: React.FC<IContestsProps> = (props: IContestsProps) => {
  return props.auditType === 'BALLOT_COMPARISON' ? (
    <ContestSelect {...props} />
  ) : (
    <ContestForm {...props} />
  )
}

export default Contests
