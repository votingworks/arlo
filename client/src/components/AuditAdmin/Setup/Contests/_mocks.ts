export const contestsInputMocks = {
  inputs: [
    { key: 'Contest Name', value: 'Contest Name' },
    { key: 'Name of Candidate/Choice 1', value: 'Choice One' },
    { key: 'Name of Candidate/Choice 2', value: 'Choice Two' },
    { key: 'Votes for Candidate/Choice 1', value: '10' },
    { key: 'Votes for Candidate/Choice 2', value: '20' },
    { key: 'Total Ballot Cards Cast for Contest', value: '30' },
  ],
  errorInputs: [
    { key: 'Contest Name', value: '', error: 'Required' },
    {
      key: 'Total Ballot Cards Cast for Contest',
      value: '',
      error:
        'Must be greater than or equal to the sum of votes for each candidate/choice',
    },
    {
      key: 'Total Ballot Cards Cast for Contest',
      value: 'test',
      error: 'Must be a number',
    },
    {
      key: 'Total Ballot Cards Cast for Contest',
      value: '-1',
      error: 'Must be a positive number',
    },
    {
      key: 'Total Ballot Cards Cast for Contest',
      value: '0.5',
      error: 'Must be an integer',
    },
    { key: 'Name of Candidate/Choice 1', value: '', error: 'Required' },
    { key: 'Name of Candidate/Choice 2', value: '', error: 'Required' },
    {
      key: 'Votes for Candidate/Choice 1',
      value: '',
      error: 'Required',
    },
    {
      key: 'Votes for Candidate/Choice 1',
      value: 'test',
      error: 'Must be a number',
    },
    {
      key: 'Votes for Candidate/Choice 1',
      value: '-1',
      error: 'Must be a positive number',
    },
    {
      key: 'Votes for Candidate/Choice 1',
      value: '0.5',
      error: 'Must be an integer',
    },
    {
      key: 'Votes for Candidate/Choice 2',
      value: '',
      error: 'Required',
    },
    {
      key: 'Votes for Candidate/Choice 2',
      value: 'test',
      error: 'Must be a number',
    },
    {
      key: 'Votes for Candidate/Choice 2',
      value: '-1',
      error: 'Must be a positive number',
    },
    {
      key: 'Votes for Candidate/Choice 2',
      value: '0.5',
      error: 'Must be an integer',
    },
  ],
}

export default contestsInputMocks
