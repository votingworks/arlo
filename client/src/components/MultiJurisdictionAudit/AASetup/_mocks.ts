import { ElementType } from '../../../types'
import { setupStages, stageTitles } from './index'
import { ISidebarMenuItem } from '../../Atoms/Sidebar'

const relativeStages = (
  stage: ElementType<typeof setupStages>,
  state: ISidebarMenuItem['state'] = 'live'
): {
  prevStage: ISidebarMenuItem
  nextStage: ISidebarMenuItem
  menuItems: ISidebarMenuItem[]
} => {
  const prevId = setupStages[setupStages.indexOf(stage) - 1]
  const prevStage: ISidebarMenuItem = {
    id: prevId,
    title: stageTitles[prevId],
    active: false,
    activate: jest.fn(),
    state,
  }
  const nextId = setupStages[setupStages.indexOf(stage) + 1]
  const nextStage: ISidebarMenuItem = {
    id: nextId,
    title: stageTitles[nextId],
    active: false,
    activate: jest.fn(),
    state,
  }
  const menuItems = setupStages.map((s: ElementType<typeof setupStages>) => ({
    id: s,
    title: stageTitles[s],
    active: s === stage,
    activate: jest.fn(),
    state,
  }))
  return { prevStage, nextStage, menuItems }
}

export default relativeStages
