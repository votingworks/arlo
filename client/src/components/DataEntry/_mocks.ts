import { IAuditBoard, BallotStatus, Interpretation } from '../../types'
import { contestMocks } from '../MultiJurisdictionAudit/Setup/Contests/_mocks'
import { IBallot } from './Ballot'

export const contest = contestMocks.filledTargeted.contests[0]

export const doneDummyBallots: { ballots: IBallot[] } = {
  ballots: [
    {
      id: 'ballot-id-1',
      batch: {
        name: '0003-04-Precinct 13 (Jonesboro Fire Department)',
        id: 'batch-id-1',
        tabulator: '11',
      },
      position: 313,
      status: BallotStatus.AUDITED,
      interpretations: [],
    },
    {
      id: 'ballot-id-2',
      batch: {
        name: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        id: 'batch-id-1',
        tabulator: '11',
      },
      position: 2112,
      status: BallotStatus.AUDITED,
      interpretations: [],
    },
    {
      id: 'ballot-id-3',
      batch: {
        name: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        id: 'batch-id-1',
        tabulator: '11',
      },
      position: 1789,
      status: BallotStatus.AUDITED,
      interpretations: [],
    },
    {
      id: 'ballot-id-4',
      batch: {
        name: '0003-04-Precinct 31 (Jonesboro Fire Department)',
        id: 'batch-id-1',
        tabulator: '11',
      },
      position: 1965,
      status: BallotStatus.NOT_FOUND,
      interpretations: [],
    },
  ],
}

export const dummyBallots: { ballots: IBallot[] } = {
  ballots: [
    {
      id: 'ballot-id-1',
      batch: {
        name: '0003-04-Precinct 13 (Jonesboro Fire Department)',
        id: 'batch-id-1',
        tabulator: null,
      },
      position: 313,
      status: BallotStatus.AUDITED,
      interpretations: [
        {
          contestId: contest.id,
          interpretation: Interpretation.VOTE,
          choiceIds: [contest.choices[0].id],
          comment: 'Good ballot',
        },
      ],
    },
    {
      id: 'ballot-id-2',
      batch: {
        name: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        id: 'batch-id-1',
        tabulator: '11',
      },
      position: 2112,
      status: BallotStatus.NOT_AUDITED,
      interpretations: [],
    },
    {
      id: 'ballot-id-3',
      batch: {
        name: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        id: 'batch-id-1',
        tabulator: '11',
      },
      position: 1789,
      status: BallotStatus.AUDITED,
      interpretations: [],
    },
    {
      id: 'ballot-id-4',
      batch: {
        name: '0003-04-Precinct 13 (Jonesboro Fire Department)',
        id: 'batch-id-2',
        tabulator: '11',
      },
      position: 313,
      status: BallotStatus.AUDITED,
      interpretations: [],
    },
    {
      id: 'ballot-id-5',
      batch: {
        name: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        id: 'batch-id-2',
        tabulator: '11',
      },
      position: 2112,
      status: BallotStatus.NOT_AUDITED,
      interpretations: [],
    },
    {
      id: 'ballot-id-6',
      batch: {
        name: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        id: 'batch-id-2',
        tabulator: '11',
      },
      position: 1789,
      status: BallotStatus.AUDITED,
      interpretations: [],
    },
    {
      id: 'ballot-id-7',
      batch: {
        name: '0003-04-Precinct 13 (Jonesboro Fire Department)',
        id: 'batch-id-3',
        tabulator: '11',
      },
      position: 313,
      status: BallotStatus.AUDITED,
      interpretations: [],
    },
    {
      id: 'ballot-id-8',
      batch: {
        name: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        id: 'batch-id-3',
        tabulator: '11',
      },
      position: 2112,
      status: BallotStatus.NOT_AUDITED,
      interpretations: [],
    },
    {
      id: 'ballot-id-9',
      batch: {
        name: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        id: 'batch-id-3',
        tabulator: '11',
      },
      position: 1789,
      status: BallotStatus.AUDITED,
      interpretations: [],
    },
    {
      id: 'ballot-id-10',
      batch: {
        name: '0003-04-Precinct 13 (Jonesboro Fire Department)',
        id: 'batch-id-4',
        tabulator: '11',
      },
      position: 313,
      status: BallotStatus.AUDITED,
      interpretations: [],
    },
    {
      id: 'ballot-id-11',
      batch: {
        name: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        id: 'batch-id-4',
        tabulator: '11',
      },
      position: 2112,
      status: BallotStatus.NOT_AUDITED,
      interpretations: [],
    },
    {
      id: 'ballot-id-12',
      batch: {
        name: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        id: 'batch-id-4',
        tabulator: '11',
      },
      position: 1789,
      status: BallotStatus.AUDITED,
      interpretations: [],
    },
    {
      id: 'ballot-id-13',
      batch: {
        name: '0003-04-Precinct 13 (Jonesboro Fire Department)',
        id: 'batch-id-5',
        tabulator: '11',
      },
      position: 313,
      status: BallotStatus.AUDITED,
      interpretations: [],
    },
    {
      id: 'ballot-id-14',
      batch: {
        name: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        id: 'batch-id-5',
        tabulator: '11',
      },
      position: 2112,
      status: BallotStatus.NOT_AUDITED,
      interpretations: [],
    },
    {
      id: 'ballot-id-15',
      batch: {
        name: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        id: 'batch-id-5',
        tabulator: '11',
      },
      position: 1789,
      status: BallotStatus.AUDITED,
      interpretations: [],
    },
    {
      id: 'ballot-id-16',
      batch: {
        name: '0003-04-Precinct 13 (Jonesboro Fire Department)',
        id: 'batch-id-6',
        tabulator: '11',
      },
      position: 313,
      status: BallotStatus.AUDITED,
      interpretations: [],
    },
    {
      id: 'ballot-id-17',
      batch: {
        name: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        id: 'batch-id-6',
        tabulator: '11',
      },
      position: 2112,
      status: BallotStatus.NOT_AUDITED,
      interpretations: [],
    },
    {
      id: 'ballot-id-18',
      batch: {
        name: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        id: 'batch-id-6',
        tabulator: '11',
      },
      position: 1789,
      status: BallotStatus.AUDITED,
      interpretations: [],
    },
    {
      id: 'ballot-id-19',
      batch: {
        name: '0003-04-Precinct 13 (Jonesboro Fire Department)',
        id: 'batch-id-7',
        tabulator: '11',
      },
      position: 313,
      status: BallotStatus.AUDITED,
      interpretations: [],
    },
    {
      id: 'ballot-id-20',
      batch: {
        name: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        id: 'batch-id-7',
        tabulator: '11',
      },
      position: 2112,
      status: BallotStatus.NOT_AUDITED,
      interpretations: [],
    },
    {
      id: 'ballot-id-21',
      batch: {
        name: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        id: 'batch-id-7',
        tabulator: '11',
      },
      position: 1789,
      status: BallotStatus.AUDITED,
      interpretations: [],
    },
    {
      id: 'ballot-id-22',
      batch: {
        name: '0003-04-Precinct 13 (Jonesboro Fire Department)',
        id: 'batch-id-8',
        tabulator: '11',
      },
      position: 313,
      status: BallotStatus.AUDITED,
      interpretations: [],
    },
    {
      id: 'ballot-id-23',
      batch: {
        name: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        id: 'batch-id-8',
        tabulator: '11',
      },
      position: 2112,
      status: BallotStatus.NOT_AUDITED,
      interpretations: [],
    },
    {
      id: 'ballot-id-24',
      batch: {
        name: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        id: 'batch-id-8',
        tabulator: '11',
      },
      position: 1789,
      status: BallotStatus.AUDITED,
      interpretations: [],
    },
    {
      id: 'ballot-id-25',
      batch: {
        name: '0003-04-Precinct 13 (Jonesboro Fire Department)',
        id: 'batch-id-9',
        tabulator: '11',
      },
      position: 313,
      status: BallotStatus.AUDITED,
      interpretations: [],
    },
    {
      id: 'ballot-id-26',
      batch: {
        name: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        id: 'batch-id-9',
        tabulator: '11',
      },
      position: 2112,
      status: BallotStatus.NOT_AUDITED,
      interpretations: [],
    },
    {
      id: 'ballot-id-27',
      batch: {
        name: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        id: 'batch-id-9',
        tabulator: '11',
      },
      position: 1789,
      status: BallotStatus.AUDITED,
      interpretations: [],
    },
  ],
}

export const dummyBoards = (): IAuditBoard[] => [
  {
    id: 'audit-board-1',
    name: 'Audit Board #1',
    jurisdictionId: 'jurisdiction-1',
    jurisdictionName: 'Jurisdiction 1',
    roundId: 'round-1',
    members: [
      {
        name: 'John Doe',
        affiliation: '',
      },
      {
        name: 'Jane Doe',
        affiliation: 'LIB',
      },
    ],
    signedOffAt: null,
  },
  {
    id: 'audit-board-2',
    name: 'Audit Board #2',
    jurisdictionId: 'jurisdiction-1',
    jurisdictionName: 'Jurisdiction 1',
    roundId: 'round-1',
    members: [],
    signedOffAt: null,
  },
]

export default {
  dummyBoards,
  dummyBallots,
}
