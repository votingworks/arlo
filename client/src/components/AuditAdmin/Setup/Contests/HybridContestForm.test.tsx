import React from 'react'
import { screen, waitFor, render } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { QueryClientProvider } from 'react-query'
import {
  withMockFetch,
  regexpEscape,
  createQueryClient,
} from '../../../testUtilities'
import Contests, { IContestsProps } from './Contests'
import { aaApiCalls } from '../../../_mocks'
import { IContest } from '../../../../types'

const hybridContestsInputMocks = {
  inputs: [
    { key: 'Name of Candidate/Choice 1', value: 'Choice One' },
    { key: 'Name of Candidate/Choice 2', value: 'Choice Two' },
    { key: 'Votes for Candidate/Choice 1', value: '10' },
    { key: 'Votes for Candidate/Choice 2', value: '20' },
  ],
  errorInputs: [
    { key: 'Name of Candidate/Choice 1', value: '', error: 'Required' },
    { key: 'Name of Candidate/Choice 2', value: '', error: 'Required' },
    {
      key: 'Votes for Candidate/Choice 1',
      value: '',
      error: 'Required',
    },
    {
      key: 'Votes for Candidate/Choice 1',
      value: 'test',
      error: 'Must be a number',
    },
    {
      key: 'Votes for Candidate/Choice 1',
      value: '-1',
      error: 'Must be a positive number',
    },
    {
      key: 'Votes for Candidate/Choice 1',
      value: '0.5',
      error: 'Must be an integer',
    },
    {
      key: 'Votes for Candidate/Choice 2',
      value: '',
      error: 'Required',
    },
    {
      key: 'Votes for Candidate/Choice 2',
      value: 'test',
      error: 'Must be a number',
    },
    {
      key: 'Votes for Candidate/Choice 2',
      value: '-1',
      error: 'Must be a positive number',
    },
    {
      key: 'Votes for Candidate/Choice 2',
      value: '0.5',
      error: 'Must be an integer',
    },
  ],
}

const renderContests = (props: Partial<IContestsProps> = {}) => {
  const goToNextStage = jest.fn()
  const goToPrevStage = jest.fn()
  return {
    goToNextStage,
    goToPrevStage,
    ...render(
      <QueryClientProvider client={createQueryClient()}>
        <Contests
          electionId="1"
          auditType="HYBRID"
          isTargeted
          goToNextStage={goToNextStage}
          goToPrevStage={goToPrevStage}
          {...props}
        />
      </QueryClientProvider>
    ),
  }
}

const apiCalls = {
  getStandardizedContests: {
    url: '/api/election/1/standardized-contests',
    response: [
      {
        name: 'Contest 1.\'"', // Make sure dots and quotes in the name work
        jurisdictionIds: ['jurisdiction-id-1', 'jurisdiction-id-2'],
      },
      {
        name: 'Contest 2',
        jurisdictionIds: ['jurisdiction-id-1'],
      },
      { name: 'Contest 3', jurisdictionIds: ['jurisdiction-id-2'] },
    ],
  },
  getContests: (contests: Omit<IContest, 'totalBallotsCast'>[]) => ({
    url: '/api/election/1/contest',
    response: { contests },
  }),
  putContests: (contests: Omit<IContest, 'totalBallotsCast'>[]) => ({
    url: '/api/election/1/contest',
    options: {
      method: 'PUT',
      body: JSON.stringify(contests),
      headers: { 'Content-Type': 'application/json' },
    },
    response: { status: 'ok' },
  }),
}

const mockUuid = jest.fn()
jest.mock('uuidv4', () => () => mockUuid())

describe('Audit Setup > Contests (Hybrid)', () => {
  let getID: () => string
  beforeEach(() => {
    // uuidMock and getID should be in sync so we can generate test data that
    // matches the UUIDs assigned when making new contests
    mockUuid.mockImplementation(
      (() => {
        let id = 0
        return () => {
          id += 1
          return id.toString()
        }
      })()
    )
    getID = (() => {
      let id = 0
      return () => {
        id += 1
        return id.toString()
      }
    })()
  })

  const newContest = () => ({
    id: getID(),
    name: 'Contest 1.\'"',
    isTargeted: true,
    numWinners: 1,
    votesAllowed: 1,
    jurisdictionIds: ['jurisdiction-id-1', 'jurisdiction-id-2'],
    choices: [
      {
        id: getID(),
        name: 'Choice One',
        numVotes: 10,
      },
      {
        id: getID(),
        name: 'Choice Two',
        numVotes: 20,
      },
    ],
  })

  it('Audit Setup > Contests', async () => {
    const expectedCalls = [
      apiCalls.getContests([]),
      aaApiCalls.getJurisdictions,
      apiCalls.getStandardizedContests,
    ]
    await withMockFetch(expectedCalls, async () => {
      const { container } = renderContests()
      await screen.findByRole('heading', { name: 'Target Contests' })
      expect(container).toMatchSnapshot()
    })
  })

  it('is able to submit form successfully', async () => {
    const contests = [newContest()]
    const expectedCalls = [
      apiCalls.getContests([]),
      aaApiCalls.getJurisdictions,
      apiCalls.getStandardizedContests,
      apiCalls.putContests(contests),
      apiCalls.getContests(contests),
    ]
    await withMockFetch(expectedCalls, async () => {
      const { goToNextStage } = renderContests()
      await screen.findByRole('heading', { name: 'Target Contests' })
      userEvent.selectOptions(
        screen.getByLabelText(/Contest Name/),
        'Contest 1.\'"'
      )
      hybridContestsInputMocks.inputs.forEach(inputData => {
        const input = screen.getByLabelText(
          new RegExp(regexpEscape(inputData.key)),
          { selector: 'input' }
        )
        userEvent.type(input, inputData.value)
      })
      expect(
        screen.queryByText('Total Ballot Cards Cast')
      ).not.toBeInTheDocument()
      userEvent.click(screen.getByRole('button', { name: 'Save & Next' }))
      await waitFor(() => expect(goToNextStage).toHaveBeenCalled())
    })
  })

  it('removes a contest', async () => {
    const contests = [
      newContest(),
      {
        ...newContest(),
        name: 'Contest 2',
        jurisdictionIds: ['jurisdiction-id-1'],
      },
      { ...newContest(), name: 'Contest 3', isTargeted: false },
    ]
    const expectedCalls = [
      apiCalls.getContests(contests),
      aaApiCalls.getJurisdictions,
      apiCalls.getStandardizedContests,
      apiCalls.putContests(contests.slice(1)),
      apiCalls.getContests(contests.slice(1)),
    ]
    await withMockFetch(expectedCalls, async () => {
      const { goToNextStage } = renderContests()
      await screen.findByRole('heading', { name: 'Target Contests' })
      userEvent.click(screen.getByRole('button', { name: 'Remove Contest 1' }))
      await waitFor(() =>
        expect(
          screen.queryByRole('heading', { name: 'Contest 1' })
        ).not.toBeInTheDocument()
      )
      userEvent.click(screen.getByRole('button', { name: 'Save & Next' }))
      await waitFor(() => expect(goToNextStage).toHaveBeenCalled())
    })
  })

  it('hides jurisdiction selection for hybrid', async () => {
    const contests = [newContest()]
    const expectedCalls = [
      apiCalls.getContests([]),
      aaApiCalls.getJurisdictions,
      apiCalls.getStandardizedContests,
      apiCalls.putContests(contests),
      apiCalls.getContests(contests),
    ]
    await withMockFetch(expectedCalls, async () => {
      const { goToNextStage } = renderContests()
      await screen.findByRole('heading', { name: 'Target Contests' })
      userEvent.selectOptions(
        screen.getByLabelText(/Contest Name/),
        'Contest 1.\'"'
      )
      hybridContestsInputMocks.inputs.forEach(inputData => {
        const input = screen.getByLabelText(
          new RegExp(regexpEscape(inputData.key)),
          {
            selector: 'input',
          }
        )
        userEvent.type(input, inputData.value)
      })
      expect(screen.queryByText('Contest Universe')).not.toBeInTheDocument()
      userEvent.click(screen.getByRole('button', { name: 'Save & Next' }))
      await waitFor(() => expect(goToNextStage).toHaveBeenCalled())
    })
  })

  it('it should not skip to next stage when targeted contest form is clean and not touched', async () => {
    const expectedCalls = [
      apiCalls.getContests([]),
      aaApiCalls.getJurisdictions,
      apiCalls.getStandardizedContests,
    ]
    await withMockFetch(expectedCalls, async () => {
      renderContests()
      await screen.findByRole('heading', { name: 'Target Contests' })
      userEvent.click(screen.getByRole('button', { name: 'Save & Next' }))
      await waitFor(() => {
        expect(screen.queryAllByText('Required').length).toBe(5)
      })
    })
  })

  it('it should skip to next stage when opportunistic contest form is clean and not touched', async () => {
    const expectedCalls = [
      apiCalls.getContests([]),
      aaApiCalls.getJurisdictions,
      apiCalls.getStandardizedContests,
    ]
    await withMockFetch(expectedCalls, async () => {
      const { goToNextStage } = renderContests({ isTargeted: false })
      await screen.findByRole('heading', { name: 'Opportunistic Contests' })
      userEvent.click(screen.getByRole('button', { name: 'Save & Next' }))
      await waitFor(() => expect(goToNextStage).toHaveBeenCalled())
    })
  })

  it('it should not skip to next stage when opportunistic contest form is touched', async () => {
    const expectedCalls = [
      apiCalls.getContests([]),
      aaApiCalls.getJurisdictions,
      apiCalls.getStandardizedContests,
    ]
    await withMockFetch(expectedCalls, async () => {
      renderContests({ isTargeted: false })

      await screen.findByRole('heading', { name: 'Opportunistic Contests' })

      userEvent.type(
        screen.getByLabelText('Votes Allowed', { selector: 'input' }),
        '2'
      )

      userEvent.click(screen.getByRole('button', { name: 'Save & Next' }))

      await waitFor(() => {
        expect(screen.queryAllByText('Required').length).toBe(5)
      })
    })
  })
})
