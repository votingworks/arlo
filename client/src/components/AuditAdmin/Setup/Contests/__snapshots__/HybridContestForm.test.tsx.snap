// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Audit Setup > Contests (Hybrid) Audit Setup > Contests 1`] = `
<div>
  <form
    data-testid="form-one"
    style="width: 100%;"
  >
    <div
      class="sc-hSdWYo hzbdZY"
    >
      <h2
        class="bp3-heading sc-iAyFgw iTWqlu"
      >
        Target Contests
      </h2>
      <div
        class="bp3-card bp3-elevation-0 sc-dqBHgY jSKysA"
        style="background: rgb(245, 248, 250);"
      >
        <div
          class="sc-bZQynM eQtLAV"
          style="margin-top: 0px;"
        >
          <h5
            class="bp3-heading"
          >
            Contest  Info
          </h5>
          <div>
            <div
              class="sc-gzVnrw kvvweo"
            >
              Select the name of the contest that will drive the audit.
            </div>
            <label
              for="contests[0].name"
            >
              Contest
               
              
               Name
              <br />
              <div
                class="bp3-html-select sc-dfVpRl khgCIW"
              >
                <select
                  field="[object Object]"
                  form="[object Object]"
                  id="contests[0].name"
                >
                  <option
                    value=""
                  />
                  <option
                    label="Contest 1.'\\""
                    value="Contest 1.'\\""
                  >
                    Contest 1.'"
                  </option>
                  <option
                    label="Contest 2"
                    value="Contest 2"
                  >
                    Contest 2
                  </option>
                  <option
                    label="Contest 3"
                    value="Contest 3"
                  >
                    Contest 3
                  </option>
                </select>
                <span
                  class="bp3-icon bp3-icon-double-caret-vertical"
                  icon="double-caret-vertical"
                >
                  <svg
                    data-icon="double-caret-vertical"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                  >
                    <desc>
                      double-caret-vertical
                    </desc>
                    <path
                      d="M5 7h6a1.003 1.003 0 00.71-1.71l-3-3C8.53 2.11 8.28 2 8 2s-.53.11-.71.29l-3 3A1.003 1.003 0 005 7zm6 2H5a1.003 1.003 0 00-.71 1.71l3 3c.18.18.43.29.71.29s.53-.11.71-.29l3-3A1.003 1.003 0 0011 9z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </span>
              </div>
            </label>
          </div>
          <div
            class="sc-gzVnrw kvvweo"
          >
            Enter the number of winners for the contest.
          </div>
          <label
            for="contests[0].numWinners"
          >
            Winners
            <div
              class="sc-gqjmRU egaEhI"
            >
              <div
                class="bp3-input-group sc-VigVT hfLuvB"
              >
                <input
                  class="bp3-input"
                  id="contests[0].numWinners"
                  name="contests[0].numWinners"
                  style="padding-right: 10px;"
                  value="1"
                />
              </div>
            </div>
          </label>
          <div
            class="sc-gzVnrw kvvweo"
          >
            Number of selections the voter can make in the contest.
          </div>
          <label
            for="contests[0].votesAllowed"
          >
            Votes Allowed
            <div
              class="sc-gqjmRU egaEhI"
            >
              <div
                class="bp3-input-group sc-VigVT hfLuvB"
              >
                <input
                  class="bp3-input"
                  id="contests[0].votesAllowed"
                  name="contests[0].votesAllowed"
                  style="padding-right: 10px;"
                  value="1"
                />
              </div>
            </div>
          </label>
        </div>
        <div
          class="sc-bZQynM eQtLAV"
        >
          <h5
            class="bp3-heading"
          >
            Candidates/Choices & Vote Totals
          </h5>
          <div
            class="sc-gzVnrw kvvweo"
          >
            Enter the name of each candidate choice that appears on the ballot for this contest.
          </div>
          <div
            class="sc-lhVmIH hSVOFV"
          >
            <div
              class="sc-bYSBpT eFidgo"
            >
              <label
                class="bp3-label sc-jtRfpW xkJJK"
              >
                Name of Candidate/Choice 
                1
                <div
                  class="sc-elJkPf kPpSXP sc-gqjmRU egaEhI"
                >
                  <div
                    class="bp3-input-group sc-VigVT hfLuvB"
                  >
                    <input
                      class="bp3-input"
                      name="contests[0].choices[0].name"
                      style="padding-right: 10px;"
                      value=""
                    />
                  </div>
                </div>
              </label>
              <label
                class="bp3-label sc-jtRfpW xkJJK"
              >
                Votes for Candidate/Choice 
                1
                <div
                  class="sc-elJkPf kPpSXP sc-gqjmRU egaEhI"
                >
                  <div
                    class="bp3-input-group sc-VigVT hfLuvB"
                  >
                    <input
                      class="bp3-input"
                      name="contests[0].choices[0].numVotes"
                      style="padding-right: 10px;"
                      value=""
                    />
                  </div>
                </div>
              </label>
            </div>
            <div
              class="sc-bYSBpT eFidgo"
            >
              <label
                class="bp3-label sc-jtRfpW xkJJK"
              >
                Name of Candidate/Choice 
                2
                <div
                  class="sc-elJkPf kPpSXP sc-gqjmRU egaEhI"
                >
                  <div
                    class="bp3-input-group sc-VigVT hfLuvB"
                  >
                    <input
                      class="bp3-input"
                      name="contests[0].choices[1].name"
                      style="padding-right: 10px;"
                      value=""
                    />
                  </div>
                </div>
              </label>
              <label
                class="bp3-label sc-jtRfpW xkJJK"
              >
                Votes for Candidate/Choice 
                2
                <div
                  class="sc-elJkPf kPpSXP sc-gqjmRU egaEhI"
                >
                  <div
                    class="bp3-input-group sc-VigVT hfLuvB"
                  >
                    <input
                      class="bp3-input"
                      name="contests[0].choices[1].numVotes"
                      style="padding-right: 10px;"
                      value=""
                    />
                  </div>
                </div>
              </label>
            </div>
            <p
              class="sc-kTUwUJ kWFRUA"
            >
              Add a new candidate/choice
            </p>
          </div>
        </div>
      </div>
      <div
        style="padding-top: 15px;"
      >
        <button
          class="bp3-button"
          type="button"
        >
          <span
            class="bp3-icon bp3-icon-add"
            icon="add"
          >
            <svg
              data-icon="add"
              height="16"
              viewBox="0 0 16 16"
              width="16"
            >
              <desc>
                add
              </desc>
              <path
                d="M10.99 6.99h-2v-2c0-.55-.45-1-1-1s-1 .45-1 1v2h-2c-.55 0-1 .45-1 1s.45 1 1 1h2v2c0 .55.45 1 1 1s1-.45 1-1v-2h2c.55 0 1-.45 1-1s-.45-1-1-1zm-3-7c-4.42 0-8 3.58-8 8s3.58 8 8 8 8-3.58 8-8-3.58-8-8-8zm0 14c-3.31 0-6-2.69-6-6s2.69-6 6-6 6 2.69 6 6-2.68 6-6 6z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
          <span
            class="bp3-button-text"
          >
            Add Contest
          </span>
        </button>
      </div>
    </div>
    <div
      class="sc-btzYZH hOMNvX"
      style="margin-top: 15px;"
    >
      <button
        class="bp3-button"
        type="button"
      >
        <span
          class="bp3-icon bp3-icon-arrow-left"
          icon="arrow-left"
        >
          <svg
            data-icon="arrow-left"
            height="16"
            viewBox="0 0 16 16"
            width="16"
          >
            <desc>
              arrow-left
            </desc>
            <path
              d="M13.99 6.99H4.41L7.7 3.7a1.003 1.003 0 00-1.42-1.42l-5 5a1.014 1.014 0 000 1.42l5 5a1.003 1.003 0 001.42-1.42L4.41 8.99H14c.55 0 1-.45 1-1s-.46-1-1.01-1z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
        <span
          class="bp3-button-text"
        >
          Back
        </span>
      </button>
      <button
        class="bp3-button bp3-intent-primary"
        type="submit"
      >
        <span
          class="bp3-button-text"
        >
          Save & Next
        </span>
        <span
          class="bp3-icon bp3-icon-arrow-right"
          icon="arrow-right"
        >
          <svg
            data-icon="arrow-right"
            height="16"
            viewBox="0 0 16 16"
            width="16"
          >
            <desc>
              arrow-right
            </desc>
            <path
              d="M14.7 7.29l-5-5a.965.965 0 00-.71-.3 1.003 1.003 0 00-.71 1.71l3.29 3.29H1.99c-.55 0-1 .45-1 1s.45 1 1 1h9.59l-3.29 3.29a1.003 1.003 0 001.42 1.42l5-5c.18-.18.29-.43.29-.71s-.12-.52-.3-.7z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
      </button>
    </div>
  </form>
</div>
`;
